src/framebuffer.ml

```diff
@@ -1,34 +1,36 @@
-
-type t = int array array
+type t = { data : int array array; mutable dirty : bool }
 
 type shader_func = int -> int
 
 type shaderi_func = int -> int -> t -> int
 
 let to_array (buffer : t) : int array array =
-  buffer
+  buffer.data
 
 let init (dimensions : int * int) (f : int -> int -> int) : t =
   let width, height = dimensions in
   if width <= 0 then raise (Invalid_argument "Invalid width");
   if height <= 0 then raise (Invalid_argument "Invalid height");
-  Array.init height (fun y ->
+  { data = Array.init height (fun y ->
     Array.init width (fun x ->
         f x y
       )
-  )
+  );
+    dirty = true }
 
 let pixel_write (x : int) (y : int) (col : int) (buffer : t) =
-  if (x >= 0) && (x < Array.length (buffer.(0))) && (y >= 0) && (y < Array.length buffer) then
-    buffer.(y).(x) <- col
+  if (x >= 0) && (x < Array.length (buffer.data.(0))) && (y >= 0) && (y < Array.length buffer.data) then
+    buffer.data.(y).(x) <- col;
+  buffer.dirty <- true
 
 let pixel_read (x : int) (y : int) (buffer : t) : int option =
-  if (x >= 0) && (x < Array.length (buffer.(0))) && (y >= 0) && (y < Array.length buffer) then
-    Some buffer.(y).(x)
+  if (x >= 0) && (x < Array.length (buffer.data.(0))) && (y >= 0) && (y < Array.length buffer.data) then
+    Some buffer.data.(y).(x)
   else
     None
 
 let draw_circle (x : int) (y : int) (r : float) (col : int) (buffer : t) =
+  buffer.dirty <- true;
```

Comment:
This calls `pixel_write`, which will set the dirty bit, so doesn't need to explicitly call it.